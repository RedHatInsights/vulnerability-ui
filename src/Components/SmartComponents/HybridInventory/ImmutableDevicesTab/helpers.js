import { getAffectedSystemsByCVE, useGetImageData } from '../../../../Helpers/APIHelper';
import { SYSTEMS_EXPOSED_HEADER } from '../../../../Helpers/constants';

export const fetchFullDeviceInfo = async (parameters, fetchImagesData) => {
    const vulnerableSystems = await getAffectedSystemsByCVE(parameters);
    const deviceIds = vulnerableSystems?.data.map(data => data.id);

    if (deviceIds.length) {
        const deviceData = await fetchImagesData({ devices_uuid: deviceIds });
        vulnerableSystems.data?.forEach(system => {
            const systemDevice = deviceData?.data?.devices.find(device => device.DeviceUUID === system.id);
            return system.attributes = { ...system.attributes, ...systemDevice };
        });
    }

    return vulnerableSystems;
};

export const useGetEntities = ({ id, setUrlParams, createRows }) => {
    const fetchImagesData = useGetImageData();
    const getEntities = async (
        _items,
        { orderBy, orderDirection, page, per_page: perPage, edgeParams, filters }
    ) => {
        const { hostnameOrId: filter } = filters;
        const sort = `${orderDirection === 'ASC' ? '' : '-'}${orderBy}`;

        const params = {
            ...edgeParams,
            page,
            page_size: perPage,
            sort,
            filter,
            ...filters?.hostGroupFilter ? {
                group_names: filters.hostGroupFilter.join(',')
            } : {}
        };

        setUrlParams?.({ ...params });

        const items = await fetchFullDeviceInfo(
            {
                ...id && { id },
                ...params
            },
            fetchImagesData
        );

        return {
            results: typeof createRows === 'function'
                ? createRows(items || {})
                : items?.data?.map(row => ({ id: row.id, ...row.attributes })),
            total: items?.meta?.total_items,
            page: items?.meta?.page,
            perPage: items?.meta?.page_size,
            meta: {
                page: items?.meta?.page,
                sort: items?.meta?.sort,
                page_size: items?.meta?.page_size,
                cves_without_errata: items?.meta?.cves_without_errata
            }
        };
    };

    return getEntities;
};

export const mergeAppColumns = (defaultColumns) => {
    const osColumn = SYSTEMS_EXPOSED_HEADER.find((column) => column.key === 'os');
    const advisoryColumn = SYSTEMS_EXPOSED_HEADER.find((column) => column.key === 'advisories_list');

    const osIndex = defaultColumns.findIndex(column =>{
        return column.key === 'system_profile';
    });
    if (osIndex !== -1) {
        defaultColumns[osIndex] = osColumn;
    }

    return [...defaultColumns, ...[advisoryColumn || []]];
};
