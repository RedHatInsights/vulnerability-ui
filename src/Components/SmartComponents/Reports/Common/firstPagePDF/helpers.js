import messages from '../../../../../Messages';

export const getAffectingFilterTexts = (systemCount) => ({
    false: `not affecting any ${systemCount}`,
    true: `impacting at least one of your ${systemCount} analyzed RHEL systems.`,
    rpmdnf: `impacting at least one of your analyzed ${systemCount} RHEL conventional systems.`,
    edge: 'impacting at least one of your analyzed ${systemCount} RHEL immutable systems.',
    'rpmdnf,edge': `impacting at least one of your analyzed ${systemCount} RHEL systems`
});

export const cveFilterStringSegments = [
    { name: 'impact', msg: messages.CVEsPDFreportFiltersSeverity, lowercase: true },
    { name: 'business_risk_id', msg: messages.CVEsPDFreportFiltersBusinessRisk, lowercase: true },
    { name: 'status_id', msg: messages.CVEsPDFreportFiltersStatus, lowercase: true },
    { name: 'rule_presence', msg: messages.CVEsPDFreportFiltersDefinedValue, lowercase: true },
    { name: 'known_exploit', msg: messages.CVEsPDFreportFiltersDefinedValue, lowercase: true },
    { name: 'advisory_available', msg: messages.CVEsPDFreportFiltersAdvisoryAvailable, lowercase: true },
    { name: 'rhel_version', msg: messages.CVEsPDFreportFiltersOsVersion }
];

export const systemsFilterStringSegments = [
    { name: 'filter', msg: messages.systemsPDFreportFiltersSearchTerm, lowercase: true },
    { name: 'rhel_version', msg: messages.systemsPDFreportFiltersOsVersion }
];
