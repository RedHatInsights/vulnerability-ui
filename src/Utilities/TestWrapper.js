import propTypes from 'prop-types';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

const TestWrapper = ({ children, store }) => {
    return (
        <IntlProvider locale="en">
            { store ?
                <Provider store={store}>
                    <Router>
                        {children}
                    </Router>
                </Provider> : children
            }
        </IntlProvider>
    );
};

TestWrapper.propTypes = {
    children: propTypes.node,
    store: propTypes.object
};

export default TestWrapper;
