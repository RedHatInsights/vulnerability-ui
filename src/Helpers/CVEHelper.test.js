import { ExternalLinkAltIcon } from '@patternfly/react-icons';
import { processDate } from '@redhat-cloud-services/frontend-components-utilities/helpers';
import Immutable from 'seamless-immutable';
import {
    createCveDetailsPage,
    createMitreLink,
    createRHDBLink,
    cveTableRowActions,
    createSystemsListTable,
    systemCveTableRowActions,
    systemExposedTableRowActions,
    systemTableRowActions
} from './CVEHelper';
import { render, screen } from '@testing-library/react';
import TestWrapper from '../Utilities/TestWrapper';

const affected_payload = Immutable({
    data: [
        {
            attributes: {
                id: 1,
                status_name: 'On-Hold',
                status_id: 2,
                cve_status_id: 3,
                rule: { value: 'testRule' },
                business_risk_id: '3'
            }
        },
        {
            attributes: {
                id: 2,
                status_name: 'On-Hold',
                status_id: 2,
                cve_status_id: 2,
                business_risk_id: '2'
            }
        }
    ],
    meta: {
        testObject: 'testObject'
    }
});

const cves_payload = Immutable({
    isLoading: false,
    payload: {
        data: {
            attributes: {
                impact: 'Important',
                public_date: '1970-01-01 00:00:00',
                description: 'testDescription',
                synopsis: 'CVE-2016-0800',
                cvss3_score: 7.6,
                cvss2_score: 6.7,
                business_risk_id: '3',
                status_id: 0
            }
        },
        meta: {
            testObject: 'testObject'
        }
    }
});

describe('CVEHelper', () => {
    it('createCveDetailsPage', () => {
        const page = createCveDetailsPage(cves_payload);
        expect(page.meta.testObject).toEqual('testObject');
        expect(page.isLoading).toEqual(false);

        expect(page.data.impact).toEqual(cves_payload.payload.data.attributes.impact);
        expect(page.data.public_date).toEqual(processDate(cves_payload.payload.data.attributes.public_date));
        expect(page.data.synopsis).toEqual(cves_payload.payload.data.attributes.synopsis);
        expect(page.data.rh_link).toEqual(createRHDBLink(cves_payload.payload.data.attributes.synopsis));
        expect(page.data.mitre_link).toEqual(createMitreLink(cves_payload.payload.data.attributes.synopsis));
        expect(page.data.cvss2_score).toEqual(cves_payload.payload.data.attributes.cvss2_score);
        expect(page.data.cvss3_score).toEqual(cves_payload.payload.data.attributes.cvss3_score);
    });

    it('createCveDetailsPage - loading', () => {
        const page = createCveDetailsPage({ isLoading: true });
        expect(page.data).toEqual([]);
        expect(page.meta).toEqual({});
        expect(page.isLoading).toEqual(true);
    });

    it('createRHDBLink - Default link label', () => {
        const { asFragment } = render(
            <TestWrapper>
                { createRHDBLink('CVE-2016-0800') }
            </TestWrapper>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('createRHDBLink - Custom link label', () => {
        let customLink = {
            id: 1,
            defaultMessage: 'Red Hat CVE database'
        };

        const { asFragment } = render(
            <TestWrapper>
                { createRHDBLink('CVE-2016-0800', customLink) }
            </TestWrapper>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('createMitreLink', () => {
        const { asFragment } = render(
            <TestWrapper>
                { createMitreLink('CVE-2016-0800') }
            </TestWrapper>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('Should createSystemsListTable return initial payload when isLoading is true', () => {
        const page = createSystemsListTable({ payload: affected_payload, isLoading: true });
        expect(page.isLoading).toEqual(true);
        expect(page.meta).toEqual({});
        expect(page.data).toEqual(affected_payload.data);
    });

    it('Should createSystemsListTable process status column', () => {
        const page = createSystemsListTable({ payload: affected_payload, isLoading: false });
        expect(page.isLoading).toEqual(false);
        expect(page.meta).toEqual(affected_payload.meta);
    });

    it('Should createSystemsListTable run without payload safely', () => {
        const page = createSystemsListTable({ isLoading: true });
        expect(page.isLoading).toEqual(true);
        expect(page.data).toEqual([]);
    });

    it('cveTableRowActions', () => {
        const testMethods = {
            showBusinessRiskModal: jest.fn(),
            showStatusModal: jest.fn()
        };
        const testBusinessObject = {
            id: 'testId',
            business_risk_id: 'estBusinessRiskID',
            business_risk_justification: 'testBusinesRiskJustification'
        };
        const testStatusObject = {
            id: 'testId',
            status_id: 'testStatusId',
            status_justification: 'testStatusJustification'
        }

        const actions = cveTableRowActions(testMethods, 0);
        actions[0].onClick(null, null, testBusinessObject);
        actions[1].onClick(null, null, testStatusObject);
        expect(actions[1].title.props.defaultMessage).toEqual('Edit status');

        expect(testMethods.showBusinessRiskModal).toHaveBeenCalledWith([
            {
                id: 'testId',
                business_risk_id: 'estBusinessRiskID',
                justification: 'testBusinesRiskJustification'
            }
        ],
        false
        );

        expect(testMethods.showStatusModal).toHaveBeenCalledWith([
            {
                id: 'testId',
                status_id: 'testStatusId',
                justification: 'testStatusJustification'
            }
        ],
        false);
    });

    it('systemCveTableRowActions ', () => {
        const testMethods = {
            showStatusModal: jest.fn()
        };
        const testStatusObject = {
            id: 'testId',
            status_id: 'testStatusId',
            cve_status_id: 'testCvestatusId',
            status_justification: 'testStatusJustification',
            cve_status_justification: 'testCveStatusJustification'
        }
        const entity = {
            display_name: 'testName'
        }

        const actions = systemCveTableRowActions(testMethods, entity, 0);
        actions[0].onClick(null, null, testStatusObject);

        expect(actions[0].title.props.defaultMessage).toEqual('Edit status');
        expect(testMethods.showStatusModal).toHaveBeenCalledWith([
            {
                id: 'testId',
                attributes: {
                    display_name: 'testName',
                    status_id: 'testStatusId',
                    cve_status_id: 'testCvestatusId',
                    status_text: 'testStatusJustification',
                    cve_status_text: 'testCveStatusJustification'
                }
            }
        ], false, { display_name: 'testName' });
    });

    it('systemExposedTableRowActions', () => {
        const showStatusModal = jest.fn();
        const testObject = {
            id: 'testId',
            status_id: 'testStatusId',
            justification: 'testJustification'
        }
        const testRowObject = {
            id: 'testId',
            status_id: 'testStatusId',
            status_text: 'testJustification'
        }
        const actions = systemExposedTableRowActions(showStatusModal, testObject);
        actions[0].onClick(null, null, testRowObject);
        expect(actions[0].title.props.defaultMessage).toEqual('Edit status');
        expect(showStatusModal).toHaveBeenCalledWith(
            [testObject],
            [testObject]
        );
    });

    it('Should systemTableRowActions display "Include system in vulnerability analysis" when system is excluded from analysis', () => {
        const optOutFunc = jest.fn();
        const rawData = { opt_out: true };
        const actions = systemTableRowActions(rawData, optOutFunc);
        expect(actions[0].title.props.defaultMessage).toEqual('Enable Vulnerability analysis on {count, plural, one {system} other {systems}}');
    });

    it('Should systemTableRowActions display "Exclude system from vulnerability analysis" when system is included in analysis', () => {
        const optOutFunc = jest.fn();
        const rawData = { opt_out: false };
        const actions = systemTableRowActions(rawData, optOutFunc);
        expect(actions[0].title.props.defaultMessage).toEqual('Disable Vulnerability analysis on {count, plural, one {system} other {systems}}');
    });

    it('Should systemTableRowActions call optOut with ID', () => {
        const optOutFunc = jest.fn();
        const rawData = { id: 'testId', display_name: 'testName', opt_out: true };
        const actions = systemTableRowActions(rawData, optOutFunc);
        actions[0].onClick(null, null, rawData);
        expect((optOutFunc)).toHaveBeenCalledWith([{"display_name": "testName", "id": "testId", "opt_out": true}], 'testName', false);
    })
});
