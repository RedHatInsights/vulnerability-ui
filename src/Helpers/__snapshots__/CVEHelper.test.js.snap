// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVEHelper createMitreLink 1`] = `
<DocumentFragment>
  <a
    href="https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-0800"
    rel="noopener noreferrer"
    target="_blank"
  >
    MITRE Database
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style="vertical-align: -0.125em;"
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
      />
    </svg>
  </a>
</DocumentFragment>
`;

exports[`CVEHelper createRHDBLink - Custom link label 1`] = `
<DocumentFragment>
  <a
    href="https://access.redhat.com/security/cve/CVE-2016-0800"
    rel="noopener noreferrer"
    target="_blank"
  >
    Red Hat CVE database
    <svg
      aria-hidden="true"
      class="pf-u-ml-sm"
      fill="currentColor"
      height="1em"
      role="img"
      style="vertical-align: -0.125em;"
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
      />
    </svg>
  </a>
</DocumentFragment>
`;

exports[`CVEHelper createRHDBLink - Default link label 1`] = `
<DocumentFragment>
  <a
    href="https://access.redhat.com/security/cve/CVE-2016-0800"
    rel="noopener noreferrer"
    target="_blank"
  >
    View in Red Hat CVE database
    <svg
      aria-hidden="true"
      class="pf-u-ml-sm"
      fill="currentColor"
      height="1em"
      role="img"
      style="vertical-align: -0.125em;"
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
      />
    </svg>
  </a>
</DocumentFragment>
`;
